### PR TITLE
Refactored type definitions

### PR DESCRIPTION
**Pull request by delvedor**
This pr improves massively our type definitions, the best way to explain how, is to show it:

Old definitions *(example taken from the old documentation)*:
```ts
import {
  Client,
  // Object that contains the type definitions of every API method
  RequestParams,
  // Interface of the generic API response
  ApiResponse,
} from '@elastic/elasticsearch'

const client = new Client({ node: 'http://localhost:9200' })

// Define the type of the body for the Search request
interface SearchBody {
  query: {
    match: { foo: string }
  }
}

// Complete definition of the Search response
interface ShardsResponse {
  total: number;
  successful: number;
  failed: number;
  skipped: number;
}

interface Explanation {
  value: number;
  description: string;
  details: Explanation[];
}

interface SearchResponse<T> {
  took: number;
  timed_out: boolean;
  _scroll_id?: string;
  _shards: ShardsResponse;
  hits: {
    total: number;
    max_score: number;
    hits: Array<{
      _index: string;
      _type: string;
      _id: string;
      _score: number;
      _source: T;
      _version?: number;
      _explanation?: Explanation;
      fields?: any;
      highlight?: any;
      inner_hits?: any;
      matched_queries?: string[];
      sort?: string[];
    }>;
  };
  aggregations?: any;
}

// Define the interface of the source object
interface Source {
  foo: string
}


// Define the search parameters
const searchParams: RequestParams.Search<SearchBody> = {
  index: 'test',
  body: {
    query: {
      match: { foo: 'bar' }
    }
  }
}

// Craft the final type definition
const response: ApiResponse<SearchResponse<Source>> = await client.search(searchParams)
console.log(response.body)
```

New definitions *(example taken from the new test)*:
```ts
// Licensed to Elasticsearch B.V under one or more agreements.
// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
// See the LICENSE file in the project root for more information

import { expectType } from 'tsd'
import { Client, ApiResponse } from '../../'

const client = new Client({
  node: 'http://localhost:9200'
})

interface SearchBody {
  query: {
    match: { foo: string }
  }
}

interface ShardsResponse {
  total: number;
  successful: number;
  failed: number;
  skipped: number;
}

interface Explanation {
  value: number;
  description: string;
  details: Explanation[];
}

interface SearchResponse<T> {
  took: number;
  timed_out: boolean;
  _scroll_id?: string;
  _shards: ShardsResponse;
  hits: {
    total: number;
    max_score: number;
    hits: Array<{
      _index: string;
      _type: string;
      _id: string;
      _score: number;
      _source: T;
      _version?: number;
      _explanation?: Explanation;
      fields?: any;
      highlight?: any;
      inner_hits?: any;
      matched_queries?: string[];
      sort?: string[];
    }>;
  };
  aggregations?: any;
}

interface Source {
  foo: string
}

// No generics
{
  const response = await client.search({
    index: 'test',
    body: {
      query: {
        match: { foo: 'bar' }
      }
    }
  })

  expectType<any>(response.body)
  expectType<any>(response.meta.context)
}

// Define only the request body
{
  const response = await client.search<SearchBody>({
    index: 'test',
    body: {
      query: {
        match: { foo: 'bar' }
      }
    }
  })

  expectType<any>(response.body)
  expectType<any>(response.meta.context)
}

// Define request body and response body
{
  const response = await client.search<SearchBody, SearchResponse<Source>>({
    index: 'test',
    body: {
      query: {
        match: { foo: 'bar' }
      }
    }
  })

  expectType<SearchResponse<Source>>(response.body)
  expectType<any>(response.meta.context)
}

// Define request body, response body and the context
{
  const response = await client.search<SearchBody, SearchResponse<Source>, string>({
    index: 'test',
    body: {
      query: {
        match: { foo: 'bar' }
      }
    }
  })

  expectType<SearchResponse<Source>>(response.body)
  expectType<string>(response.meta.context)
}
```

**TL;DR**
Now every API makes better use of the generics and overloading, so you can (or not, by default request/response bodies are `any`) define the request/response bodies in the generics.
```ts
// request and response bodies are `any`
client.search(...)
// request body is `SearchBody` and response body is `any`
client.search<SearchBody>(...)
// request body is `SearchBody` and response body is `SearchResponse`
client.search<SearchBody, SearchResponse>(...)
```

This *should* not be a breaking change, as every generics defaults to `any`. It might happen to some users that the code breaks, but our test didn't detect any of it, probably because they were not robust enough. However, given the gigantic improvement in the developer experience, we have decided to release this change in the 7.x line.
This pr also introduces a vastly improved testing infrastructure for addressing the old fragile tests.
This pr should also help the future work we will do to address https://github.com/elastic/elasticsearch-js/issues/933.

Don't get scared by the size of the pr, most of the code (in `index.d.ts`) is generated from a script, once you have reviewed a few lines, you have reviewed them all! ;)

Thank you to @fox1t for helping me designing this!

### CODE REVIEW

**Comment by villasv** [comment]
This is sooo good. Thanks so much for the hard work :-)

File: scripts/utils/generateMain.js [hunk 1]
```diff
@@ -160,4 +164,66 @@ function isSnakeCased (str) {
...
+        { key: `${camelify(api)}<T = any, B = any, C = any> (params: RequestParams.${Name}<T>, options: TransportRequestOptions)`, val: `Promise<ApiResponse<B, C>>` },
+        { key: `${camelify(api)}<T = any, B = any, C = any> (callback: callbackFn<B, C>)`, val: `TransportRequestCallback` },
+        { key: `${camelify(api)}<T = any, B = any, C = any> (params: RequestParams.${Name}<T>, callback: callbackFn<B, C>)`, val: `TransportRequestCallback` },
+        { key: `${camelify(api)}<T = any, B = any, C = any> (params: RequestParams.${Name}<T>, options: TransportRequestOptions, callback: callbackFn<B, C>)`, val: `TransportRequestCallback` }
+      ])
+    }
+    return methods
+  } else {
+    let methods = [
+      { key: `${api}<B = any, C = any> ()`, val: `Promise<ApiResponse<B, C>>` },
```
[the remark this answers is not in the document]
Comment (delvedor) [reply scripts/utils/generateMain.js]: The client supports both snaked_cased apis and camelCased apis, for instance the latest async search will be available both as `async_search` and `asyncSearch`.
That if check only verifies if we should or not generate the camelCased api as well.

File: scripts/utils/generateMain.js [hunk 2]
```diff
@@ -160,4 +164,66 @@ function isSnakeCased (str) {
   return !!~str.indexOf('_')
 }
 
+function buildMethodDefinition (api, name, body) {
+  const Name = name[0].toUpperCase() + name.slice(1)
+  if (body) {
```
[the remark this answers is not in the document]
Comment (delvedor) [reply scripts/utils/generateMain.js]: I'm not sure I follow, what do you mean?

File: scripts/utils/generateMain.js [hunk 3]
```diff
@@ -160,4 +164,66 @@ function isSnakeCased (str) {
...
+        { key: `${camelify(api)}<T = any, B = any, C = any> (params: RequestParams.${Name}<T>, callback: callbackFn<B, C>)`, val: `TransportRequestCallback` },
+        { key: `${camelify(api)}<T = any, B = any, C = any> (params: RequestParams.${Name}<T>, options: TransportRequestOptions, callback: callbackFn<B, C>)`, val: `TransportRequestCallback` }
+      ])
+    }
+    return methods
+  } else {
+    let methods = [
+      { key: `${api}<B = any, C = any> ()`, val: `Promise<ApiResponse<B, C>>` },
+      { key: `${api}<B = any, C = any> (params: RequestParams.${Name})`, val: `Promise<ApiResponse<B, C>>` },
+      { key: `${api}<B = any, C = any> (params: RequestParams.${Name}, options: TransportRequestOptions)`, val: `Promise<ApiResponse<B, C>>` },
```
[the remark this answers is not in the document]
Comment (delvedor) [reply scripts/utils/generateMain.js]: Definitely!

File: test/types/errors.test-d.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+{
+  const err = new errors.ElasticsearchClientError()
+  expectType<string>(err.name)
+  expectType<string>(err.message)
+}
+
+{
+  const err = new errors.TimeoutError('message', response)
+  expectType<string>(err.name)
```
[the remark this answers is not in the document]
Comment (delvedor) [reply test/types/errors.test-d.ts]: Absolutely yes, as we don't have any other way to check if our changes will cause or not a breaking change, having the test also for the type definitions is the best solution for solving this issue.

The use of the `tsd` package is becoming quite standard in the community, I'm using it already in other projects successfully.

File: scripts/utils/generateMain.js [hunk 5]
```diff
@@ -7,10 +7,11 @@
 'use strict'
 
 const { readdirSync } = require('fs')
+const { join } = require('path')
 const dedent = require('dedent')
 const deepmerge = require('deepmerge')
 
-function genFactory (folder) {
+function genFactory (folder, paths) {
```
[the remark this answers is not in the document]
Comment (delvedor) [reply scripts/utils/generateMain.js]: Currently, I'm executing that manually and we don't have a mechanism for detecting changes in Elasticsearch. The easiest way to figure out if the client is not updated is that the CI will go red.
As clients team, we are working on a solution that will address this issue :)

**Comment by delvedor** [comment]
> I tried to run node scripts/generate --tag=v7.6.1 but it didn't finish after 30 min. Are there any specific setup requirements to run the command locally?

Nope, it should just work. Did it log something?
As first step (which usually is the one that takes longer), it will clone locally the elasticsearch repository, maybe something went wrong there?

File: test/types/errors.test-d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (delvedor) [reply test/types/errors.test-d.ts]: As an example on why testing the type definitions is useful, I've forgotten to update the new naming and defaults ina couple of places, and the type test started failing :)

File: scripts/utils/generateMain.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (restrry) [reply scripts/utils/generateMain.js]: `body` type is not `string`, but `boolean` here, so the argument name could be more inlined with the type.

**Comment by restrry** [comment]
> Nope, it should just work. Did it log something?

It clones elasticsearch repo  in `elasticsearch-js/elasticsearch` folder, but it tries to checkout `elasticsearch-js` repo on `v7.6.1` tag instead of `elasticsearch`, so `elasticsearch` folder remains empty.

File: scripts/utils/generateMain.js [hunk 6]
```diff
@@ -160,4 +164,70 @@ function isSnakeCased (str) {
...
+        { key: `${camelify(api)}<RequestBody = any, Response = any, Context = unknown>(params: RequestParams.${Name}<RequestBody>, options: TransportRequestOptions)`, val: `Promise<ApiResponse<Response, Context>>` },
+        { key: `${camelify(api)}<RequestBody = any, Response = any, Context = unknown>(callback: callbackFn<Response, Context>)`, val: `TransportRequestCallback` },
+        { key: `${camelify(api)}<RequestBody = any, Response = any, Context = unknown>(params: RequestParams.${Name}<RequestBody>, callback: callbackFn<Response, Context>)`, val: `TransportRequestCallback` },
+        { key: `${camelify(api)}<RequestBody = any, Response = any, Context = unknown>(params: RequestParams.${Name}<RequestBody>, options: TransportRequestOptions, callback: callbackFn<Response, Context>)`, val: `TransportRequestCallback` }
+      ])
+    }
+    return methods
+  } else {
+    let methods = [
+      { key: `${api}<Response = any, Context = unknown>()`, val: `Promise<ApiResponse<Response, Context>>` },
```
Comment (restrry) [inline scripts/utils/generateMain.js]: why Response is `any`, Context is `unknown`?

File: scripts/utils/generateMain.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (restrry) [reply scripts/utils/generateMain.js]: just to confirm: not all `snake_case` API methods have `camelCase` counterparts?

File: scripts/utils/generateMain.js (hunk 6 above)
Comment (delvedor) [reply scripts/utils/generateMain.js]: Because I'm trying to avoid any breaking change for current users. The current default value of generics is `any`, while it has never been set for the context, so I've decided to use `unknown`, which is more appropriate.

**Comment by delvedor** [comment]
> It clones elasticsearch repo in elasticsearch-js/elasticsearch folder, but it tries to checkout elasticsearch-js repo on v7.6.1 tag instead of elasticsearch, so elasticsearch folder remains empty.

I fear there is a bug and the very first run of the code generation should use master, I've just run in a freshly cloned repo
```
node scripts/generate --branch master
```
and everything worked as expected.

File: scripts/utils/generateMain.js (hunk 6 above)
Comment (delvedor) [reply scripts/utils/generateMain.js]: I've just updated the code to improve the request body definition, it should only be `Record<string, any> | string | Buffer | ReadableStream`.
In the future, we should update the `Record<string, any>` with the real interface.

Comment (delvedor) [reply scripts/utils/generateMain.js]: I'm thinking of defining the `BodyType` as follows:
```ts
type BodyType<T = Record<string, any>> = T | string | Buffer | ReadableStream
```
In this way, we should be able to add the correct interfaces gradually.
I'm not sure if it's fine, but ts is not complaining :D

File: lib/Transport.d.ts [hunk 7]
```diff
@@ -66,31 +70,30 @@ export interface RequestEvent<T = any, C = any> {
...
 // ApiResponse and RequestEvent are the same thing
 // we are doing this for have more clear names
-export interface ApiResponse<T = any, C = any> extends RequestEvent<T, C> {}
+export interface ApiResponse<Response = any, Context = unknown> extends RequestEvent<Response, Context> {}
 
-declare type anyObject = {
-  [key: string]: any;
-};
+export type BodyType = Record<string, any> | string | Buffer | ReadableStream
+export type NDBodyType = Record<string, any>[] | string[] | Buffer | ReadableStream
```
Comment (restrry) [inline lib/Transport.d.ts]: Don't see tests for `BodyType`: Buffer & ReadableStream & `NDBodyType`

Comment (delvedor) [reply lib/Transport.d.ts]: They are coming ;)
I'm fixing another issue that I've discovered while writing the test! :D

**Comment by delvedor** [comment]
I did another change, hopefully, the last one!
Before a method was defined as follows:
```ts
search<RequestBody = BodyType, Response = any, Context = unknown>(params: RequestParams.Search<RequestBody>, options: TransportRequestOptions): Promise<ApiResponse<Response, Context>>
```
Which worked well if the user was passing their own types, but if they were using the defaults, ts didn't "followed" the `BodyType`, because given it's a default value and it is used in the method signature, ts was overriding  it with the user-provided value,  making the following code valid:
```ts
 // note that number is not present in BodyType
client.search({ index: 'test', body: 42 })
```
For addressing this issue, now we are using `extends` in the `RequestBody` definition.
```ts
search<RequestBody extends BodyType, Response = any, Context = unknown>(params: RequestParams.Search<RequestBody>, options: TransportRequestOptions): Promise<ApiResponse<Response, Context>>
```
And as result the following code is no longer valid.
```ts
client.search({ index: 'test', body: 42 })
```

Furthermore, I've used `RequestBody` also as default value for `Response`. This might be a breaking change for the users, but they were checking in any case the returned type, now it is only more specific.

Finally, `BodyType` and `NDBodyType` accept a generic, which we will use in the future when we will start adding the request/response bodies definitions.
The code will look like this:
```ts
search<RequestBody extends BodyType<SearchBody>, Response = BodyType<SearchResponse>, Context = unknown>(params: RequestParams.Search<RequestBody>, options: TransportRequestOptions): Promise<ApiResponse<Response, Context>>
```

I fear we should accept the fact that we could release breaking changes for the type definitions in minors, as ts is a rapidly evolving language and they are releasing breaking changes in minors as well.

File: lib/Transport.d.ts [hunk 8]
```diff
@@ -70,10 +70,10 @@ export interface RequestEvent<Response = any, Context = unknown> {
 
 // ApiResponse and RequestEvent are the same thing
 // we are doing this for have more clear names
-export interface ApiResponse<Response = any, Context = unknown> extends RequestEvent<Response, Context> {}
+export interface ApiResponse<Response = BodyType, Context = unknown> extends RequestEvent<Response, Context> {}
 
-export type BodyType = Record<string, any> | string | Buffer | ReadableStream
-export type NDBodyType = Record<string, any>[] | string[] | Buffer | ReadableStream
+export type BodyType<T = Record<string, any>> = T | string | Buffer | ReadableStream
```
Comment (Mpdreamz) [inline lib/Transport.d.ts]: 👍 💯

File: index.d.ts [hunk 9]
```diff
@@ -106,628 +100,3416 @@ interface ClientOptions {
   }
 }
 
-declare class Client extends EventEmitter {
+declare class Client<RequestBody = any, Response = any, Context = unknown> extends EventEmitter {
```
Comment (Mpdreamz) [inline index.d.ts]: These generics on the class should not be needed since each method (rightfully) declares its own generics

File: index.d.ts [hunk 10]
```diff
@@ -106,628 +100,3416 @@ interface ClientOptions {
...
   extend: ClientExtends;
-  child(opts?: ClientOptions): Client;
+  child(opts?: ClientOptions): Client<RequestBody, Response, Context>;
   close(callback?: Function): Promise<void> | void;
   /* GENERATED */
   async_search: {
-    delete: ApiMethod<RequestParams.AsyncSearchDelete>
-    get: ApiMethod<RequestParams.AsyncSearchGet>
-    submit: ApiMethod<RequestParams.AsyncSearchSubmit>
+    delete<Response = BodyType, Context = unknown>(): Promise<ApiResponse<Response, Context>>
```
Comment (Mpdreamz) [inline index.d.ts]: Personally a fan of the following naming rule for generics:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/generic-type-naming.md

This makes them stick out more e.g `TResponse` `TContext` and `TRequest`.

File: scripts/utils/generateMain.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (delvedor) [reply scripts/utils/generateMain.js]: All `snake_case` API methods have a `camelCase` counterpart :)
